### PR TITLE
1233: Implement lists as actual list

### PR DESCRIPTION
**Pull request by lunars97**
### Short description

Implement lists as actual list for accessibility according to **Prüfbericht
zur Barrierefreiheit**

### Proposed changes

- Searchresults
- “Integreat deiner Region vor!”

### Resolved issues

Fixes: #2167 

---

### CODE REVIEW

**Review by steffenkleinle — CHANGES_REQUESTED** [review]
This issue is for both web and native and I think there is quite a few places where we don't use lists. Probably we shouldn't use lists in all these cases, but just examples (all found by searching for .map() functions in the code):
* native: Categories.jsx -> Footer
* native: CalendarChoiceModal
* native: CategoryListItem -> rendering of sub category list items
* native: Selector
* native: OpeningHours
* native: PoiFilters
* web: CitySelector
* web: OpeningHours
* web: PoiFilters
* web: SearchPage
* web: Selector

So probably we should split up the issue and do some more research in which cases it makes sense and in which cases it does not make sense.

File: native/src/routes/CityNotCooperating.tsx [hunk 1]
```diff
@@ -53,30 +57,51 @@ const StepExplanation = styled.Text`
...
+  explanation: string
+}
+
 const CityNotCooperating = (): ReactElement | null => {
   const { t } = useTranslation('cityNotCooperating')
   const [isCopied, setIsCopied] = useState<boolean>(false)
   const template = buildConfig().featureFlags.cityNotCooperatingTemplate
   const CityNotCooperatingIcon = buildConfigAssets().CityNotCooperatingIcon
 
+  const steps: Step[] = [
```
[the remark this answers is not in the document]
Comment (LeandraH) [reply native/src/routes/CityNotCooperating.tsx:92]: I'd also think that it's a bit much but it is one of the things that has been explicitly criticized in the report from Pfennigparade.

File: native/src/components/SearchListItem.tsx [hunk 2]
```diff
@@ -100,6 +102,8 @@ const SearchListItem = ({
       </DirectionContainer>
     </FlexStyledLink>
   )
+
+  return <List items={category} renderItem={({ item: category }) => category} Header={SearchList} />
```
[the remark this answers is not in the document]
Comment (LeandraH) [reply native/src/components/SearchListItem.tsx:106]: The search results have been named as an example of where we didn't implement a list properly, so there are still things we need to do here. Might be worth looking at it with a screenreader.

File: native/src/routes/CityNotCooperating.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (steffenkleinle) [reply native/src/routes/CityNotCooperating.tsx:92]: This exact example was criticized? I think we should discuss this, we shouldn't just do everything if we don't think it makes sense.

[the remark this answers is not in the document]
Comment (lunars97) [reply native/src/routes/CityNotCooperating.tsx:92]: @LeandraH @steffenkleinle  should I leave it for now as it is?

[the remark this answers is not in the document]
Comment (steffenkleinle) [reply native/src/routes/CityNotCooperating.tsx:92]: Yes, lets wait for more opinions here. Maybe we should discuss this in a call. Sorry for that!

**Comment by LeandraH** [comment]
Let's have a chat about this in Monday's Weekly?

File: native/src/routes/CityNotCooperating.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (ztefanie) [reply native/src/routes/CityNotCooperating.tsx:92]: Yes this exact example was mentioned. I am fine with keeping it as if is or with changing it to an actual list.

File: native/src/components/SearchListItem.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (lunars97) [reply native/src/components/SearchListItem.tsx]: I am currently inserting FlatList to native: **Categories.jsx -> Footer.** Should I open for it another issue or can I push changes here? I am a bit confused right now

[the remark this answers is not in the document]
Comment (lunars97) [reply native/src/components/SearchListItem.tsx]: @LeandraH SearchResults are working good with a screen reader. Should I leave it then?

[the remark this answers is not in the document]
Comment (LeandraH) [reply native/src/components/SearchListItem.tsx]: Wait, so you couldn't reproduce the issues named in the report? In https://nextcloud.tuerantuer.org/index.php/apps/files/?dir=/2023-10%20Aktion%20Mensch_Barrierefreiheit%20f%C3%BCr%20alle_bis%20zu%20350k&openfile=4955578 under "Listenartige Elemente sind fehlerhaft implementiert"?

[the remark this answers is not in the document]
Comment (LeandraH) [reply native/src/components/SearchListItem.tsx]: Why don't we keep this PR for native changes and open another issue for checking over the web stuff?

[the remark this answers is not in the document]
Comment (lunars97) [reply native/src/components/SearchListItem.tsx]: @LeandraH It says that we need to include announcement to search results.

[the remark this answers is not in the document]
Comment (LeandraH) [reply native/src/components/SearchListItem.tsx]: Yes, announcements of how long the list is, which list entry we are focusing on, and that we've reached the end of the list. Did I understand correctly that you tested these with a screen reader on the search results and found all of them?

[the remark this answers is not in the document]
Comment (lunars97) [reply native/src/components/SearchListItem.tsx]: @LeandraH no, there is no announcement of list length. I have been looking for a solution

[the remark this answers is not in the document]
Comment (lunars97) [reply native/src/components/SearchListItem.tsx]: Can I for now implement all cases you mentioned above @steffenkleinle here in this PR?

[the remark this answers is not in the document]
Comment (lunars97) [reply native/src/components/SearchListItem.tsx]: @LeandraH seems like it should announce automatically. But so far I was not able to reproduce it with emulators

[the remark this answers is not in the document]
Comment (LeandraH) [reply native/src/components/SearchListItem.tsx]: That's too bad! I don't think it makes sense to implement the other cases until we know how to do that properly, though. Do you have any other ideas how to get the announcements in there?

[the remark this answers is not in the document]
Comment (LeandraH) [reply native/src/components/SearchListItem.tsx]: Thank you :)

File: native/src/routes/SearchModal.tsx [hunk 3]
```diff
@@ -102,6 +102,7 @@ const SearchModal = ({
       contentWithoutHtml={item.contentWithoutHtml}
       language={languageCode}
       query={query}
+      accessibilityHint={t(`It navigates to the description of the item`)}
```
Comment (ztefanie) [inline native/src/routes/SearchModal.tsx:105]: This should be moved to our translations.json file, so the users get this output in their language.
Also i think "description" is not really the correct term, maybe "page" is better.
Better place to put this, is in the `SearchListItem.tsx` as all searchListItems have the same hint, no need to pass it around with props.

File: native/src/routes/CityNotCooperating.tsx [hunk 4]
```diff
@@ -3,19 +3,22 @@ import React, { ReactElement, useState } from 'react'
 import { useTranslation } from 'react-i18next'
```
Comment (ztefanie) [inline native/src/routes/CityNotCooperating.tsx:3]: Please revert this file.

File: web/src/routes/CityNotCooperatingPage.tsx [hunk 5]
```diff
@@ -13,7 +13,7 @@ const Container = styled.div`
   flex-direction: column;
```
Comment (ztefanie) [inline web/src/routes/CityNotCooperatingPage.tsx:13]: Please revert this file.

File: native/src/components/base/TextButton.tsx [hunk 6]
```diff
@@ -25,12 +25,31 @@ type TextButtonProps = {
   text: string
   onPress: () => Promise<void> | void
   disabled?: boolean
+  accessible?: boolean
+  focusable?: boolean
+  accessibilityRole?: AccessibilityRole | undefined
```
Comment (ztefanie) [inline native/src/components/base/TextButton.tsx:30]: No need to have the `undefined` here, as this prop is optional, because of the question mark, the `undefined` type is added automatically.

Also there is no need to have the `accessibilityRole` as a prop here, as we are in a Button component, the accessibilityRole should always be "button" here. So just add below `accessibilityRole='button'` and remove the option to pass an other prop.

File: native/src/components/base/TextButton.tsx [hunk 7]
```diff
@@ -25,12 +25,31 @@ type TextButtonProps = {
   text: string
   onPress: () => Promise<void> | void
   disabled?: boolean
+  accessible?: boolean
+  focusable?: boolean
```
Comment (ztefanie) [inline native/src/components/base/TextButton.tsx:29]: Please remove the `accessible` and `focusable` props from this component, as you do not use them and they are set by default for every `Pressable`, see the React Native docs: https://reactnative.dev/docs/accessibility#accessible

File: native/src/components/__tests__/SearchListItem.spec.tsx [hunk 8]
```diff
@@ -14,6 +14,7 @@ jest.mock('react-native-webview', () => ({
 describe('SearchListItem', () => {
```
Comment (ztefanie) [inline native/src/components/__tests__/SearchListItem.spec.tsx:14]: This file can be reverted, when you move the accessibilityHint inside the SearchListItem.tsx. See my comment in `SearchModal.tsx`

File: native/src/components/SearchListItem.tsx [hunk 9]
```diff
@@ -46,6 +46,7 @@ type SearchListItemProps = {
   contentWithoutHtml: string
```
Comment (ztefanie) [inline native/src/components/SearchListItem.tsx:46]: This file can be reverted, see my comment on the file `SearchModal.tsx`

File: native/src/components/List.tsx [hunk 10]
```diff
@@ -11,10 +11,12 @@ const NoItemsMessage = styled.Text`
 
 type ListProps<T> = {
   items: Array<T>
-  noItemsMessage: ReactElement | string
+  noItemsMessage?: ReactElement | string
```
Comment (ztefanie) [inline native/src/components/List.tsx:14]: Why have you changed this to be optional?

File: native/src/components/List.tsx [hunk 11]
```diff
@@ -11,10 +11,12 @@ const NoItemsMessage = styled.Text`
...
   renderItem: (props: { item: T; index: number }) => ReactElement
   Header?: ReactElement
   Footer?: ReactElement
+  scrollEnabled?: boolean
+  accessibilityRole?: AccessibilityRole | undefined
```
Comment (ztefanie) [inline native/src/components/List.tsx:19]: No need to pass the `accessibilityRole` here as a prop. This component has a FlatList inside, which should always have the `accessibiltyRole="grid"`, see [RN docs](https://reactnative.dev/docs/accessibility#accessibilityrole). This will also add the "the in/out of grid announcements to the android GridView.", e.g. "four elements in the list"

File: native/src/components/List.tsx [hunk 12]
```diff
@@ -45,6 +49,8 @@ const List = <T,>({
       paddingHorizontal: 10,
     }}
     onEndReachedThreshold={1}
+    scrollEnabled={scrollEnabled}
+    accessibilityRole={accessibilityRole}
```
Comment (ztefanie) [inline native/src/components/List.tsx:53]: Just set it to the value `"grid"` here.

File: native/src/components/LayoutedScrollView.tsx [hunk 13]
```diff
@@ -7,18 +7,20 @@ import Layout from './Layout'
 type LayoutedScrollViewProps = {
   children?: React.ReactNode
   refreshControl?: React.ReactElement<RefreshControlProps>
+  nestedScrollEnabled?: boolean
```
Comment (ztefanie) [inline native/src/components/LayoutedScrollView.tsx:10]: Why is this change needed?

File: native/src/components/CategoryListItem.tsx [hunk 14]
```diff
@@ -82,15 +83,19 @@ const CategoryListItem = ({
...
-    {subCategories.map(subCategory => (
-      <SubCategoryListItem
-        key={subCategory.path}
-        subCategory={subCategory}
-        resourceCache={resourceCache[subCategory.path] ?? {}}
-        onItemPress={onItemPress}
-        language={language}
-      />
-    ))}
+    <List
```
Comment (ztefanie) [inline native/src/components/CategoryListItem.tsx:86]: Maybe you can rename this file and component to `CategoryList.tsx`, as this component does not render one listItem, as the name says, but the whole list. This was the case before your changes, but maybe you can nevertheless adjust it.

**Review by ztefanie — CHANGES_REQUESTED** [review]
As Steffen suggested, it is a little over-engineered to introduce a list for only the two items int the "CityNotCooperation"-View, so please revert these two files. Sorry about that, it was originally mentioned in the ticket, but sometimes we realize the best solution, not until we review.

File: native/src/components/LayoutedScrollView.tsx (hunk 13 above)
Comment (lunars97) [reply native/src/components/LayoutedScrollView.tsx:10]: @ztefanie while there was a virtual list error saying that `VirtualizedLists should never be nested inside plain ScrollViews with the same orientation because it can break windowing and other functionality`

File: native/src/components/List.tsx (hunk 10 above)
Comment (lunars97) [reply native/src/components/List.tsx:14]: because there is always a ts error saying that this prop is missing or should I include it everywhere?

File: native/src/routes/SearchModal.tsx (hunk 3 above)
Comment (lunars97) [reply native/src/routes/SearchModal.tsx:105]: @ztefanie if I put it inside `SearchListItem` it causes error

File: native/src/components/__tests__/SearchListItem.spec.tsx (hunk 8 above)
Comment (lunars97) [reply native/src/components/__tests__/SearchListItem.spec.tsx:14]: @ztefanie I was unable to insert it into SearchListItem. It caused an undefined error

File: native/src/routes/SearchModal.tsx (hunk 3 above)
Comment (ztefanie) [reply native/src/routes/SearchModal.tsx]: Not sure which error you get, but here is the solution i suggest: https://github.com/digitalfabrik/integreat-app/pull/2583/files

Comment (lunars97) [reply native/src/routes/SearchModal.tsx]: @ztefanie I also did the same thing, let me try it again :)

**Review by ztefanie — APPROVED** [review]
Looks good :tada: Nice work. 

Tested on Android, works fine.

If you fix something we mentioned in a comment, you can click "resolve conversation", so it is easy to see, which conversations are open and need e.g. an answer and which one are "done".

Also please squash the commits into one commit.

It is really nice to see, how you never give up and always make your PRs nice and functioning code. Well done :clap:

File: native/src/components/LayoutedScrollView.tsx [hunk 15]
```diff
@@ -15,6 +15,7 @@ const LayoutedScrollView = (props: LayoutedScrollViewProps): ReactElement => {
     <Layout>
       <ScrollView
         keyboardShouldPersistTaps='always'
+        nestedScrollEnabled
```
Comment (steffenkleinle) [inline native/src/components/LayoutedScrollView.tsx:18]: why do we need that?

File: native/src/routes/SearchModal.tsx [hunk 16]
```diff
@@ -113,6 +114,7 @@ const SearchModal = ({
         <List
           items={searchResults}
           renderItem={renderItem}
+          accessibilityLabel={`${searchCategories(categories, query).length} ${t('searchResults')}`}
```
Comment (steffenkleinle) [inline native/src/routes/SearchModal.tsx:117]: This should go to translations.json with proper plural handling and using a placeholder for the number.

File: web/src/routes/CityNotCooperatingPage.tsx [hunk 17]
```diff
@@ -119,5 +119,4 @@ const CityNotCooperatingPage = ({ languageCode }: CityNotCooperatingPageProps):
     </Layout>
   )
 }
-
```
Comment (steffenkleinle) [inline web/src/routes/CityNotCooperatingPage.tsx:122]: Please readd

File: native/src/components/SearchListItem.tsx [hunk 18]
```diff
@@ -57,6 +59,7 @@ const SearchListItem = ({
   contentWithoutHtml,
   onItemPress,
   query,
+  t,
```
Comment (steffenkleinle) [inline native/src/components/SearchListItem.tsx:62]: Instead of passing this you should better use the useTranslation hook.

File: translations/translations.json [hunk 19]
```diff
@@ -8401,6 +8401,8 @@
     "de": {
       "nothingFound": "Es wurden leider keine passenden Ergebnisse gefunden.",
       "searchPlaceholder": "Suche",
+      "itemHint": "Es navigiert zur Beschreibung der Seite",
```
Comment (steffenkleinle) [inline translations/translations.json:8404]: Optional: Not sure about this but I think something like that would be better:
```suggestion
      "itemHint": "Suchergebnis öffnen",
```

File: native/src/components/CategoryListItem.tsx (hunk 14 above)
Comment (steffenkleinle) [reply native/src/components/CategoryListItem.tsx:86]: @ztefanie I think this is not correct. This renders one list item which, in turn, includes a list of sub category items. Therefore the name should stay the way it is.

File: native/src/components/LayoutedScrollView.tsx (hunk 15 above)
Comment (lunars97) [reply native/src/components/LayoutedScrollView.tsx:18]: @steffenkleinle yep, this exact error

Comment (steffenkleinle) [reply native/src/components/LayoutedScrollView.tsx:18]: But do you have a resource that says that this `nestedScrollEnabled` fixes the error? Or did you just try things until you found a solution?

Comment (lunars97) [reply native/src/components/LayoutedScrollView.tsx:18]: @steffenkleinle many developers mentioned this solution when I was searching for an answer why

Comment (lunars97) [reply native/src/components/LayoutedScrollView.tsx:18]: [Here is the link](https://github.com/facebook/react-native/issues/31697)

File: native/src/components/Categories.tsx [hunk 20]
```diff
@@ -81,7 +81,7 @@ const Categories = ({
       path={category.path}
       AfterContent={category.organization && <OrganizationContentInfo organization={category.organization} />}
       Footer={children.map(it => (
```
Comment (steffenkleinle) [inline native/src/components/Categories.tsx:83]: So in theory we have a list of lists here, there is multiple categories that are mapped to `SubCategoryList`s. In theory the childrens should also be rendered in a list. Can also be done in a separate issue though.

File: native/src/components/SubCategoryList.tsx [hunk 21]
```diff
@@ -52,35 +52,35 @@ export const CategoryThumbnail = styled(SimpleImage)<{ language: string }>`
...
   subCategories,
   resourceCache,
   onItemPress,
-}: CategoryListItemProps): ReactElement => (
+}: SubCategoryListProps): ReactElement => (
   <>
     <FlexStyledLink onPress={() => onItemPress({ path: category.path })}>
       <DirectionContainer language={language}>
-        <CategoryEntryContainer>
+        <SubCategoryEntryContainer>
```
Comment (steffenkleinle) [inline native/src/components/SubCategoryList.tsx:73]: Sorry but these renamings are wrong :see_no_evil: The sub categories are only rendered below (lines 86 until 98). If I think more about it, the name of this should actually stay `CategoryListItem`. The component renders a category list item. The fact that a part of that is that it renders sub categories is just an implementation detail and not really relevant to the usage... Sorry for the confusion, my mistake!

File: translations/translations.json [hunk 22]
```diff
@@ -8412,8 +8412,9 @@
     "de": {
       "nothingFound": "Es wurden leider keine passenden Ergebnisse gefunden.",
       "searchPlaceholder": "Suche",
-      "itemHint": "Es navigiert zur Beschreibung der Seite",
-      "searchResults": "Suchergebnisse",
+      "itemHint": "Suchergebnis öffnen",
+      "searchResult": "{{count}} Suchergebnis",
+      "searchResult_other": "{{count}} Suchergebnisse",
```
[the remark this answers is not in the document]
Comment (steffenkleinle) [reply translations/translations.json:8417]: I think if we want to support that we should add all the suffixes in all the german and englisch translations like this:
``` json
  "searchResultsCount_zero": "0 Suchergebnisse",
  "searchResultsCount_one": "1 Suchergebnis",
  "searchResultsCount_two": "2 Suchergebnisse",
  "searchResultsCount_few": "{{count}} Suchergebnisse",
  "searchResultsCount_many": "{{count}} Suchergebnisse",
  "searchResultsCount_other": "{{count}} Suchergebnisse"
```
And for most languages they will all be the same, just for arabic and a few others translators than have to change some.

File: native/src/components/Categories.tsx (hunk 20 above)
Comment (steffenkleinle) [reply native/src/components/Categories.tsx:83]: If you don't want to do it in this PR we can just open a new issue for this :)

Comment (lunars97) [reply native/src/components/Categories.tsx]: @steffenkleinle I implemented it :)

**Review by steffenkleinle — APPROVED** [review]
Tested on android, well done!
